### DISCUSSION TITLE
apt upgrade stuck at "Setting up libc6:amd64 (2.37-14) ..."

### DISCUSSION THREAD

**Discussion by liyafe1997** [microsoft/WSL#11097] [General] [answered]
### Windows Version

10.0.22631.3085

### WSL Version

2.0.15.0

### Are you using WSL 1 or WSL 2?

- [X] WSL 2
- [ ] WSL 1

### Kernel Version

5.15.133.1-1

### Distro Version

Debian testing

### Repro Steps

In Debian Testing

```
sudo apt update
sudo apt dist-upgrade
```

While upgrading `libc6:amd64 (2.37-14)`, it stuck forever. 

### Expected Behavior

Should be able to upgrade

### Actual Behavior

![screenshot](https://github.com/microsoft/WSL/assets/18359157/efbb678a-cbca-445f-a690-9753099f24f0)

And I can see `telinit u` takes 100% CPU

A lot of people also having this problem:
https://stackoverflow.com/questions/77885938/stuck-setting-up-libc6i386

https://unix.stackexchange.com/questions/767340/apt-stuck-on-setting-up-libc6

**Comment by cotste** [comment, marked as answer]
I had the same issue, https://github.com/microsoft/WSL/discussions/11035#discussioncomment-8149890 fixed it.

As far as I can tell if you don't have 

`[boot]
systemd=true`

The /run/systemd/system directory is not created and the tests in the postinst for libc fail:
```bash
    # Restart init.  Currently handles chroots, systemd and upstart, and
    # assumes anything else is going to not fail at behaving like
    # sysvinit:
    TELINIT=yes
    if ischroot 2>/dev/null; then
        # Don't bother trying to re-exec init from a chroot:
        TELINIT=no
    elif [ -n "${DPKG_ROOT:-}" ]; then
        # Do not re-exec init if we are operating on a chroot from outside:
        TELINIT=no
    elif [ -d /run/systemd/system ]; then
        # Restart systemd on upgrade, but carefully.
        # The restart is wanted because of LP: #1942276 and Bug: #993821
        # The care is needed because of https://bugs.debian.org/753725
        # (if systemd --help fails the system might still be quite broken but
        # that seems better than the kernel panic that results if systemd
        # cannot reexec itself).
        TELINIT=no
        if systemd --help >/dev/null 2>/dev/null; then
            systemctl daemon-reexec
        else
            echo "Error: Could not restart systemd, systemd binary not working" >&2
        fi
    fi
    if [ "$TELINIT" = "yes" ]; then
        telinit u 2>/dev/null || true ; sleep 1
    fi
```

**Comment by tj64** [comment]
The above fix did not work for me, but this answer did: 
[mtraceur](https://stackoverflow.com/users/4372452/mtraceur)
 [Jan 23 at 2:53](https://stackoverflow.com/questions/77781232/apt-stucked-in-setting-up-libc6#comment137269112_77781232)
"_I fixed this with sudo mv /usr/sbin/telinit /usr/sbin/telinit.bak and then sudo ln -s /usr/bin/true /usr/sbin/telinit_"

But then with the next upgrade to "Setting up libc6:amd64 (2.37-15) ..." my fix apparently was overriden again in the process, and the upgrade got stuck forever again. 

```
$ ls -al /usr/sbin/ |grep telinit
lrwxrwxrwx  1 root root        16 Jan 29 11:46 telinit -> ../bin/systemctl
lrwxrwxrwx  1 root root        16 Nov 22 23:23 telinit.bak -> ../bin/systemctl
```

So I had to repeat the fix
```
sudo mv /usr/sbin/telinit /usr/sbin/telinit.bak2
sudo ln -s /usr/bin/true /usr/sbin/telinit

```
and I could do this again
`sudo dpgk --configure -a`

to successfully do this in the end 
`sudo apt upgrade`

**Reply by jellopudding** [reply]
I used a similar method that worked. adding `[boot] systemd=true` to `/etc/wsl.conf` did not work for my case as well.

**Reply by Tuk9ski** [reply]
Tks brother, u saved me

**Reply by jackby03** [reply]
You're f*ck god men

**Reply by db92-coder** [reply]
If I could up vote this answer multiple times I would. I was just about at the point of smashing everything because nothing would work and this worked perfect straight away. Thanks again

**Reply by JanKotekFactoryPal** [reply]
you saved me 🙏

**Reply by zararashigal** [reply]
Lovely, thank you very much! This helped me continue with the process. ^^

**Reply by elhabib44** [reply]
thank you very much

**Reply by achryrn** [reply]
thank you!

**Reply by tj64** [reply]
replying to myself:  on this bugreport [DEBIAN bug=1060387](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1060387) it's mention that this should be fixed: 

> This was a systemd bug fixed in version 255.4-1.

For me, it did not happen again, so I checked my systemd version
``systemctl --version
` systemd` 255 (255.4-1+b1)

Conclusion: upgrading systemd will fix this, if you can, otherwise wait for the upgrade of the distro and apply this workaround repeatedly until then.

**Reply by nucleare** [reply]
I tried a lot of other things before this, but this is what finally did it, thank you.

**Reply by tempchuks** [reply]
this is awesome worked for me

**Reply by CallMePlayer** [reply]
Doing `sudo apt update` and then `sudo apt install --only-upgrade systemd` then `sudo apt upgrade` worked for my Kali Linux WSL install

**Reply by AdityaMitra5102** [reply]
This worked. Thanks. Also the 'dpkg' spelling is wrong in your answer, please fix it.

**Reply by fathiahabsyi** [reply]
Thanks works for me!

**Reply by xsswon** [reply]
It works for me, thank you😊.

**Reply by myariffx** [reply]
thanks. it works for me.

**Reply by SlyDcooper** [reply]
worked for me too thanks bro

**Reply by elrashedy1992** [reply]
worked for

**Reply by raiden076** [reply]
🫡

**Reply by MukhtaarAziz** [reply]
It is Worked For Me.

**Reply by inno-waluza** [reply]
it worked for me

**Reply by DavidParrillas** [reply]
thanks a lot bro

**Reply by ya01in** [reply]
worked for me on kali distro, thx~

**Reply by photosurf** [reply]
this worked thank you!!

**Reply by stevenkyalo** [reply]
Totally works...thanks

**Reply by Mubasshir873** [reply]
thanks bro

**Reply by wan-04** [reply]
this worked for me

**Reply by ggk570** [reply]
Worked for me. Thanks

**Reply by Aprea123** [reply]
Worked for me. Thanks

**Reply by AliMaazKhan** [reply]
Life saviour!!

**Reply by mfrancisbeehive** [reply]
This worked for me as well!

**Reply by r4gnir** [reply]
it worked for me after :3 , thank you.

**Reply by LaserGage** [reply]
This was it! Thank you :)

**Reply by issam-hub** [reply]
legend legend legend

**Reply by jon-bee** [reply]
saved my life

**Reply by alexsmartv** [reply]
Worked everything well, i was stuck with the upgrade on kali 2024 physical install and wsl2.

**Reply by blankuserrr** [reply]
> The above fix did not work for me, but this answer did: [mtraceur](https://stackoverflow.com/users/4372452/mtraceur) [Jan 23 at 2:53](https://stackoverflow.com/questions/77781232/apt-stucked-in-setting-up-libc6#comment137269112_77781232) "_I fixed this with sudo mv /usr/sbin/telinit /usr/sbin/telinit.bak and then sudo ln -s /usr/bin/true /usr/sbin/telinit_"
> 
> But then with the next upgrade to "Setting up libc6:amd64 (2.37-15) ..." my fix apparently was overriden again in the process, and the upgrade got stuck forever again.
> 
> ```
> $ ls -al /usr/sbin/ |grep telinit
> lrwxrwxrwx  1 root root        16 Jan 29 11:46 telinit -> ../bin/systemctl
> lrwxrwxrwx  1 root root        16 Nov 22 23:23 telinit.bak -> ../bin/systemctl
> ```
> 
> So I had to repeat the fix
> 
> ```
> sudo mv /usr/sbin/telinit /usr/sbin/telinit.bak2
> sudo ln -s /usr/bin/true /usr/sbin/telinit
> ```
> 
> and I could do this again `sudo dpgk --configure -a`
> 
> to successfully do this in the end `sudo apt upgrade`

i love you

**Reply by imbas007** [reply]
work thank

**Comment by lanaparadinha** [comment]
> and I could do this again
> `sudo dpgk --configure -a`

sudo dpkg --configure -a  << correctly

Thx this work for me too

**Comment by ya01in** [comment]
one liner
```bash
sudo mv /usr/sbin/telinit /usr/sbin/telinit.bak2; sudo ln -s /usr/bin/true /usr/sbin/telinit; sudo dpkg --configure -a && sudo apt upgrade
```

**Reply by fierceduck001** [reply]
great. this worked. thank you so much.

**Reply by cisamu123** [reply]
**Thanks!**

**Reply by alvnz11** [reply]
Thankss

**Reply by rohitbaijal** [reply]
Thanks!

**Reply by saikoostark** [reply]
you are legend

**Reply by EuxyL** [reply]
THANKS ALOT !!
